src/replication/replication_manager.cpp

```diff
@@ -0,0 +1,273 @@
+          ReplicaConnect(replica_name, replica_hostname, replica_port);
+        }
+        break;
+      default:
+        NOISEPAGE_ASSERT(false, "Impossible.");
+        break;
+    }
+  }
+  hosts_file.close();
+}
+
+void ReplicationManager::ReplicaConnect(const std::string &replica_name, const std::string &hostname, uint16_t port) {
+  replicas_.try_emplace(replica_name, messenger_, replica_name, hostname, port);
+}
+
+void ReplicationManager::ReplicaSend(const std::string &replica_name, const ReplicationManager::MessageType type,
+                                     const std::string &msg, bool block) {
+  if (!replication_enabled_) {
+    REPLICATION_LOG_WARN(fmt::format("Skipping send -> {} as replication is disabled."));
+    return;
+  }
+
+  REPLICATION_LOG_TRACE(fmt::format("Send -> {} (type {} block {}): msg size {}", replica_name,
+                                    static_cast<uint8_t>(type), block, msg.size()));
+  bool completed = false;
+  try {
+    messenger_->SendMessage(
+        GetReplicaConnection(replica_name), msg,
+        [this, block, &completed](common::ManagedPointer<messenger::Messenger> messenger,
+                                  const messenger::ZmqMessage &msg) {
+          if (block) {
+            // If this isn't a blocking send, then completed will fall out of scope.
+            completed = true;
+            blocking_send_cvar_.notify_all();
+          }
+        },
+        static_cast<uint64_t>(type));
+
+    if (block) {
+      std::unique_lock<std::mutex> lock(blocking_send_mutex_);
+      // If the caller requested to block until the operation was completed, the thread waits.
+      blocking_send_cvar_.wait(lock, [&completed] { return completed; });
+      lock.unlock();
+    }
+  } catch (const MessengerException &e) {
+    REPLICATION_LOG_WARN(fmt::format("[FAILED] Send -> {} (type {} block {}): msg size {}", replica_name,
+                                     static_cast<uint8_t>(type), block, msg.size()));
+  }
+}
+
+void ReplicationManager::ReplicateBuffer(storage::BufferedLogWriter *buffer) {
+  if (!replication_enabled_) {
+    REPLICATION_LOG_WARN(fmt::format("Skipping replicate buffer as replication is disabled."));
+    return;
+  }
+
+  NOISEPAGE_ASSERT(buffer != nullptr,
+                   "Don't try to replicate null buffers. That's pointless."
+                   "You might plausibly want to track statistics at some point, but that should not happen here.");
+  // TODO(WAN): Is it true that a replica will never want to send its buffers to the primary?
+  // TODO(WAN): Cache the identity check into a bool?
+  if (identity_ == "primary") {
+    common::json j;
+    // TODO(WAN): Add a size and checksum to message.
+    j["buf_id"] = next_buffer_sent_id_++;
+    j["content"] = nlohmann::json::to_cbor(std::string(buffer->buffer_, buffer->buffer_size_));
+
+    for (const auto &replica : replicas_) {
+      ReplicaSend(replica.first, MessageType::REPLICATE_BUFFER, j.dump(), true);
```

Comment:
the next goal is to figure out why everything catches on fire when `true` is flipped to `false`

Additionally, I don't think we want the `LogSerializerTask` thread running this code... ideally the buffer is handed off to the replication manager's queue and something executes this later. @jkosh44 thoughts on where this belongs?